dask/dataframe/core.py

```diff
@@ -2330,6 +2335,128 @@ def quantile(df, q):
+
+    If the number of partitions isn't changing, then the ratio of fixed
+    percentiles to random percentiles is 2 to 1.  If repartitioning goes from
+    a very high number of partitions to a very low number of partitions, then
+    we use more random percentiles, because a stochastic approach will be more
+    stable to potential correlations in the data that may cause a few equally-
+    spaced partitions to under-sample the data.
+
+    The more partitions there are, then the more total percentiles will get
+    calculated across all partitions.  Squaring the number of partitions
+    approximately doubles the number of total percentiles calculates, so
+    num_total_percentiles ~ sqrt(num_partitions).  The number of percentiles
+    calculated on a single partition is proportional to the relative length
+    of the partition.  This should provide adequate resolution and allow the
+    number of partitions to scale.
+
+    For numeric data, one could instead use T-Digest for floats and Q-Digest
+    for ints to calculate approximate percentiles.  Our current method works
+    for any dtype.
+
+    """
+    # *waves hands*
+    random_percentage = 1 / (1 + (4.0 * num_new / num_old)**0.5)
+    num_percentiles = num_new * (num_old + 22)**0.55 * chunk_length / total_length
+    num_fixed = int(num_percentiles * (1 - random_percentage)) + 2
+    num_random = int(num_percentiles * random_percentage) + 2
+
+    if num_fixed + num_random + 2 >= chunk_length:
+        return np.linspace(0, 100, chunk_length + 1)
+
+    q_fixed = np.linspace(0, 100, num_fixed)
+    q_random = np.random.rand(num_random) * 100
+    q = np.concatenate([q_fixed, q_random])
+    q.sort()
+    return q
+
+
+def _repartition_quantiles(df, npartitions):
+    """ Approximate quantiles of Series used for repartitioning
+    """
+    assert isinstance(df, Series)
+    from dask.array.percentile import _percentile, merge_percentiles
+
+    q = np.linspace(0, 1, npartitions + 1)
+    # currently, only Series has quantile method
+    # Index.quantile(list-like) must be pd.Series, not pd.Index
+    df_name = df.name
+    merge_type = lambda v: pd.Series(v, index=q, name=df_name)
+    return_type = df._constructor
+    if issubclass(return_type, Index):
+        return_type = Series
+
+    # pandas uses quantile in [0, 1]
+    # numpy / everyone else uses [0, 100]
+    qs = np.asarray(q) * 100
+    token = tokenize(df, qs)
+
+    if len(qs) == 0:
+        name = 're-quantiles-' + token
+        empty_index = pd.Index([], dtype=float)
+        return Series({(name, 0): pd.Series([], name=df.name, index=empty_index)},
+                       name, df.name, [None, None])
+    else:
+        new_divisions = [0.0, 1.0]
+
+    name1 = 're-quantiles-1-' + token
+    len_dsk = dict(((name1, i), (len, key)) for i, key in enumerate(df._keys()))
+
+    total_name = 're-quantiles-len-' + token
+    total_dsk = {total_name: (sum, sorted(len_dsk))}
```

Comment:
Knowing the length is only important if there is high variance in the lengths of partitions (more specifically, if most of the values are in a very small percentage of partitions).  We can assume approximately-uniform partition length if you prefer to avoid a full block.